MelonLoader.Installer/InstallerUtils.cs

```diff
@@ -1,19 +1,40 @@
 ﻿using System.IO.Compression;
+using System.Security.Cryptography;
 
 namespace MelonLoader.Installer;
 
 public static class InstallerUtils
 {
     public static HttpClient Http { get; }
+    private static readonly SHA512 Hasher = SHA512.Create();
 
     static InstallerUtils()
     {
         Http = new();
         Http.DefaultRequestHeaders.Add("User-Agent", $"MelonLoader Installer v{Program.VersionName}");
     }
 
-    public static async Task<string?> DownloadFileAsync(string url, Stream destination, InstallProgressEventHandler? onProgress)
+    private static async Task<string?> FetchFile(string url, Stream destination, bool useCache, InstallProgressEventHandler? onProgress)
     {
+        // Cache preparation
+        var parentDirectory = Path.GetFileName(Path.GetDirectoryName(url)) ?? "";
+        var fileCache = Path.Combine(Config.CacheDir, "Cache", parentDirectory, Path.GetFileName(url));
```

Comment:
This assumes the file name is unique for every download, which already breaks everything. ML versions do not have unique file names per version, so only one version will be cached.